### DISCUSSION TITLE
Turbo streams like API

### DISCUSSION THREAD

**Discussion by thetutlage** [unpoly/unpoly#440] [Ideas]
I have been using Unpoly on almost every web app I create. The experience has been superb, except I am missing the ability to define DOM actions from the server.

Right now, every Unpoly action results in updating some part of the DOM using the `up-target` attribute. However, there are cases, when I want to delete some element or append to a list of collection.

[Turbo streams](https://turbo.hotwired.dev/reference/streams) API is designed exactly to do that. Instead of reacting to events on the frontend, you can let the backend decide how to modify the DOM after a given action.

For example:

- I have a list of items and each item has a `DELETE` button next to it. 
- Clicking the `DELETE` button triggers the server request.
- Server responds with a turbo frame of action `remove`.
- And the item gets removed from the DOM.

**Comment by triskweline** [comment]
In your example I would re-render the entire list after deleting item, reusing the template I already have.

While I don't personally prefer the style of RJS / TurboStreams you could build something similar yourself using the [`X-Up-Events`](https://unpoly.com/X-Up-Events) header.

E.g. the server sends:

```http
X-Up-Events: [{ type: 'element:destroy', selector: '#list-item123' }]
X-Up-Target: :none
```

Your frontend picks up the `element:destroy` event and removes a matching fragment:

```js
up.on('element:destroy', function({ selector }) {
  let fragment = up.fragment.get(selector)
  if (fragment) fragment.destroy()
})
```

**Reply by thetutlage** [reply]
Thanks for sharing the events idea. I will opt for it.

> In your example I would re-render the entire list after deleting item, reusing the template I already have.

Just sharing: I think, re-rendering a list can lead to bad UX in many scenarios. For example, my current viewport has list items, `3 - 2 - 1`, I deleted the item `2` and refreshed the list. In meantime, the server got 20 new entries, and the item `3` in the UI update shifted to next page (considering pagination).

The app I am working on, we do not shift the UI unless user takes an explicit action. If you modify/delete a list item and the server notices that the list items have changed, it will send some markup saying `20 new items have been added. Refresh`. Just like the way twitter does.
